[libc++][concepts] P2404R3: Move-only types for equality_comparable_with, totally_ordered_with, and three_way_comparable_with

randomnetcat opened the pull request:
This implements all of [P2404R3](https://wg21.link/p2404r3)'s concept changes.

Fixes #105210.

randomnetcat committed:
P2404R3: Move-only types for relational concepts

Zingam commented on libcxx/include/__compare/three_way_comparable.h:
```diff
@@ -37,6 +38,20 @@ concept three_way_comparable =
 
+#  if _LIBCPP_STD_VER >= 23
+
+template <class _Tp, class _Up, class _cat = partial_ordering>
```
```suggestion
template <class _Tp, class _Up, class _Cat = partial_ordering>
```
I believe we use capital letter for template parameters.

randomnetcat replied:
Yep, sorry. Fixed.

randomnetcat committed:
Fix _Cat capitalization

randomnetcat committed:
Make test class naming consistent

Zingam commented on libcxx/test/std/language.support/cmp/cmp.concept/three_way_comparable_with.compile.pass.cpp:
```diff
@@ -223,4 +224,31 @@ struct SpaceshipNonConstArgument {
+static_assert(!check_three_way_comparable_with<MoveOnlyIntComparable, int>());
+#endif // TEST_STD_VER >= 23
+
+struct ImmobileIntComparable {
```
Do you mean `NonMovable`? For consistency:

```suggestion
struct NonMovableIntComparable{
```

`Movable` vs `NonMovable` is what is used in the tests suite mostrly.

Zingam commented on libcxx/include/__concepts/comparison_common_type.h:
```diff
@@ -0,0 +1,39 @@
+
+#if _LIBCPP_STD_VER >= 23
+
+template <class _Tp, class _Up, class _C = common_reference_t<const _Tp&, const _Up&>>
```
```suggestion
template <class _Tp, class _Up, class _Cp = common_reference_t<const _Tp&, const _Up&>>
```

I wonder if we need to have the `p` letter for consistency too. 

BTW AFAIK the latest tendency is to prefer human readable names sometimes.

Zingam commented:
The link to the paper in the description of the PR doesn't seem to be a valid one.

randomnetcat committed:
immobile -> nonmovable

randomnetcat committed:
_C -> _CommonRef

randomnetcat commented:
Fixed, sorry.

randomnetcat committed:
Update other comparison tests

cjdb commented on libcxx/include/__concepts/equality_comparable.h:
```diff
@@ -37,6 +38,22 @@ concept __weakly_equality_comparable_with =
 
+#  if _LIBCPP_STD_VER >= 23
+
+// clang-format off
+template <class _Tp, class _Up>
+concept equality_comparable_with =
+    equality_comparable<_Tp> && equality_comparable<_Up> &&
+    __comparison_common_type_with<_Tp, _Up> &&
+    equality_comparable<
+        common_reference_t<
+            __make_const_lvalue_ref<_Tp>,
+            __make_const_lvalue_ref<_Up>>> &&
+    __weakly_equality_comparable_with<_Tp, _Up>;
+// clang-format on
```
Same here.

cjdb commented on libcxx/include/__compare/three_way_comparable.h:
```diff
@@ -37,6 +38,20 @@ concept three_way_comparable =
 
+#  if _LIBCPP_STD_VER >= 23
+
+template <class _Tp, class _Up, class _Cat = partial_ordering>
+concept three_way_comparable_with =
+    three_way_comparable<_Tp, _Cat> && three_way_comparable<_Up, _Cat> && __comparison_common_type_with<_Tp, _Up> &&
+    three_way_comparable<common_reference_t<__make_const_lvalue_ref<_Tp>, __make_const_lvalue_ref<_Up>>, _Cat> &&
+    __weakly_equality_comparable_with<_Tp, _Up> && __partially_ordered_with<_Tp, _Up> &&
+    requires(__make_const_lvalue_ref<_Tp> __t, __make_const_lvalue_ref<_Up> __u) {
+      { __t <=> __u } -> __compares_as<_Cat>;
+      { __u <=> __t } -> __compares_as<_Cat>;
+    };
```
Rather than spelling the definition out twice, can you make `__common_comparison_type_with` available in C++20, with the value of `true`, please?

randomnetcat replied:
Would a better way to do that be to define it as `common_reference_with<__make_const_lvalue_ref<_Tp>, __make_const_lvalue_ref<_Up>>` in 20 so that there's only one "the common comparison type exists" concept?

cjdb commented on libcxx/include/__compare/three_way_comparable.h:
```diff
@@ -37,6 +38,20 @@ concept three_way_comparable =
       { __a <=> __b } -> __compares_as<_Cat>;
     };
 
+#  if _LIBCPP_STD_VER >= 23
```
I'd prefer we invert this, so that we have "before C++23" and "since C++23". Had to strain a little to realise that this was since/before. Same for other files, please :)

cjdb reviewed (commented):
Thanks for working on this!

I'm inclined to suggest that we treat this as if it were a defect report, and backport this to C++20, so that we have a canonical definition of these concepts across standards. I think that will, in the long term, be more beneficial to users than having the concept change between standards.

randomnetcat committed:
Invert test conditionals

randomnetcat committed:
Don't split definitions of equality_comparable_with/totally_ordered_with

randomnetcat commented:
> I'm inclined to suggest that we treat this as if it were a defect report, and backport this to C++20, so that we have a canonical definition of these concepts across standards. I think that will, in the long term, be more beneficial to users than having the concept change between standards.

Addressed the review comments, but haven't done this. I'm not sure if you meant "please do this" or "this is something we could consider".

cjdb commented:
> > I'm inclined to suggest that we treat this as if it were a defect report, and backport this to C++20, so that we have a canonical definition of these concepts across standards. I think that will, in the long term, be more beneficial to users than having the concept change between standards.
> 
> I'm not sure if you meant "please do this" or "this is something we could consider".

Definitely the latter. We typically don't make unilateral decisions, and doing it for something like this would definitely be not okay.

randomnetcat committed:
Fix last >= 23 check

Zingam commented on libcxx/include/__concepts/comparison_common_type.h:
```diff
@@ -0,0 +1,51 @@
+#ifndef _LIBCPP___CONCEPTS_COMPARISON_COMMON_TYPE_H
+#define _LIBCPP___CONCEPTS_COMPARISON_COMMON_TYPE_H
+
+#include "__concepts/common_reference_with.h"
```
```suggestion
#include <__concepts/common_reference_with.h>
```

Elsewhere also

frederick-vs-ja commented:
> I'm inclined to suggest that we treat this as if it were a defect report, and backport this to C++20, so that we have a canonical definition of these concepts across standards. I think that will, in the long term, be more beneficial to users than having the concept change between standards.

Currently, only MSVC STL has implemented this paper and leaves it C++23-only. I'd like to ask what do implementors/maintainers of other implementations think "as usual". (I'm personally neutral.)

CC @JMazurkiewicz @StephanTLavavej @jwakely

ldionne commented on libcxx/test/std/algorithms/alg.sorting/alg.sort/partial.sort.copy/ranges_partial_sort_copy.pass.cpp:
```diff
@@ -84,6 +89,12 @@ struct NoComparator {};
 
+#if TEST_STD_VER < 23
+static_assert(!HasPartialSortCopyIter<int*, int*, MoveOnly*, MoveOnly*>);
+#else
+static_assert(HasPartialSortCopyIter<int*, int*, MoveOnly*, MoveOnly*>);
```
```suggestion
#else
// P2404
static_assert(HasPartialSortCopyIter<int*, int*, MoveOnly*, MoveOnly*>);
```

ldionne commented on libcxx/test/std/algorithms/alg.sorting/alg.sort/partial.sort.copy/ranges_partial_sort_copy.pass.cpp:
```diff
@@ -109,14 +120,20 @@ static_assert(!HasPartialSortCopyRange<R<int*>, RandomAccessRangeNotDerivedFrom>
 
+#if TEST_STD_VER < 23
+static_assert(!HasPartialSortCopyRange<R<int*>, R<MoveOnly*>>);
+#else
+static_assert(HasPartialSortCopyRange<R<int*>, R<MoveOnly*>>);
```
```suggestion
#else
// P2404
static_assert(HasPartialSortCopyRange<R<int*>, R<MoveOnly*>>);
```

ldionne commented on libcxx/test/std/concepts/concepts.compare/concept.equalitycomparable/equality_comparable_with.compile.pass.cpp:
```diff
@@ -1119,4 +1119,14 @@ static_assert(
+#if TEST_STD_VER < 23
+static_assert(!check_equality_comparable_with<move_only_equality_with_int, int>());
+static_assert(!check_equality_comparable_with<std::unique_ptr<int>, std::nullptr_t>());
+#else
+static_assert(check_equality_comparable_with<move_only_equality_with_int, int>());
+static_assert(check_equality_comparable_with<std::unique_ptr<int>, std::nullptr_t>());
```
```suggestion
#else
// P2404
static_assert(check_equality_comparable_with<move_only_equality_with_int, int>());
static_assert(check_equality_comparable_with<std::unique_ptr<int>, std::nullptr_t>());
```

ldionne commented on libcxx/test/std/language.support/cmp/cmp.concept/three_way_comparable_with.compile.pass.cpp:
```diff
@@ -223,4 +224,31 @@ struct SpaceshipNonConstArgument {
+
+#if TEST_STD_VER < 23
+static_assert(!check_three_way_comparable_with<MoveOnlyIntComparable, int>());
+#else
+static_assert(check_three_way_comparable_with<MoveOnlyIntComparable, int>());
```
```suggestion
#else
// P2404
static_assert(check_three_way_comparable_with<MoveOnlyIntComparable, int>());
```

ldionne commented on libcxx/test/support/compare_types.h:
```diff
@@ -529,4 +529,30 @@ struct ForwardingTestObject {
   constexpr bool operator>=(const ForwardingTestObject&) const& { return false; }
```
Comment not attached to this line. I think we're missing an update to the `__cpp_lib_concepts` feature-test macro, the paper seems to mention something.

randomnetcat replied:
I wasn't sure if there were any other changes that that feature test macro might indicate, so I didn't want to update it and potentially promise something that wasn't implemented.

randomnetcat replied:
Alright, it changed at https://github.com/cplusplus/draft/pull/5701/files#diff-f46371f476893d323cfcb1383e9a70f07ff1c1f0c217e22294d0bf937aae6794 from the value that libc++ had previously, so this seems safe.

ldionne reviewed (commented):
Thanks for your first contribution! This looks pretty good, but I have a few comments.

In terms of applying this as a DR, I would be OK with doing that if the other implementers also do it.

randomnetcat committed:
Fix quotation mark includes

randomnetcat committed:
Add comments to tests

StephanTLavavej commented:
I would be fine with changing MSVC's STL to align with making this a de facto DR, if libstdc++ is also on-board. This doesn't take any new publicly-visible names, and it seems like a highly compatible change (Annex C lists a theoretical break but you have to be looking pretty hard for it).

randomnetcat commented:
So... should I be looking at implementing this without the version check?

ldionne commented:
@jwakely Do you think it makes sense to apply P2404R3 as a DR or do you see this as a real feature?

libc++ and MSVC STL are fine with a DR.

cjdb commented:
> So... should I be looking at implementing this without the version check?

Let's hold off until we get an answer from Jonathan, but assuming libstdc++ consents, yes please.

ldionne commented:
@randomnetcat Let's rebase this onto `main` and implement it as a DR (i.e. no version check). When you're ready, please ping me and I'll commit some time to it so we can get it across the finish line.

randomnetcat committed:
Bump __cpp_lib_concepts

randomnetcat committed:
Implement P2404 as a DR rather than only in 23

frederick-vs-ja commented on libcxx/include/__concepts/comparison_common_type.h:
```diff
@@ -0,0 +1,41 @@
+#include <__concepts/same_as.h>
+#include <__config>
+#include <__type_traits/common_reference.h>
+#include <__type_traits/make_const_lvalue_ref.h>
```
```suggestion
#include <__type_traits/make_const_lvalue_ref.h>
#include <__type_traits/remove_cvref.h>
```

randomnetcat committed:
Add missing include

Co-authored-by: A. Jiang <de34@live.cn>

randomnetcat committed:
Remove unused includes

ldionne committed:
Merge branch 'main' into p2404

jwakely commented:
Likewise for GCC 16: https://github.com/gcc-mirror/gcc/commit/03562c1e023d01cacf099a40842e617d27148c74

ldionne committed:
Merge branch 'main' into p2404

ldionne committed:
Conditionalize on Clang

ldionne reviewed (approved):
LGTM with passing CI, @frederick-vs-ja do you see anything else that needs changing before this lands?

frederick-vs-ja committed:
Tweak workaround and update links

frederick-vs-ja committed:
Update entry for P2404R3 in Cxx23Papers.csv

frederick-vs-ja committed:
Add entry for P2404R3 to 22.rst
